File: scalechain-script/src/main/scala/io/scalechain/blockchain/script/HashCalculator.scala
package io.scalechain.blockchain.script

import io.scalechain.blockchain.proto.{Hash, BlockHeader, Block, Transaction}
import io.scalechain.blockchain.proto.codec.{BlockHeaderCodec, BlockCodec, TransactionCodec}

object HashCalculator {
  def transactionHash(transaction : Transaction ) : Hash = {
    val serializedBytes = TransactionCodec.serialize(transaction)

    // Run SHA256 twice and reverse bytes.
    val hash = io.scalechain.crypto.HashFunctions.hash256( serializedBytes )

    Hash( hash.value.reverse )
  }

  def blockHeaderHash(blockheader:BlockHeader) : Hash = {
    val serializedBlockHeader = BlockHeaderCodec.serialize(blockheader)

    // Run SHA256 twice and reverse bytes.
    val blockHeaderHash = io.scalechain.crypto.HashFunctions.hash256( serializedBlockHeader )

    Hash( blockHeaderHash.value.reverse )
  }
}

File: scalechain-script/src/main/kotlin/io/scalechain/blockchain/script/HashCalculator.kt
package io.scalechain.blockchain.script

import io.netty.buffer.ByteBuf
import io.netty.buffer.Unpooled
import io.scalechain.blockchain.proto.Hash
import io.scalechain.blockchain.proto.BlockHeader
import io.scalechain.blockchain.proto.Block
import io.scalechain.blockchain.proto.Transaction
import io.scalechain.blockchain.proto.codec.BlockHeaderCodec
import io.scalechain.blockchain.proto.codec.BlockCodec
import io.scalechain.blockchain.proto.codec.TransactionCodec
import io.scalechain.blockchain.proto.codec.CodecInputOutputStream
import io.scalechain.crypto.HashFunctions
import io.scalechain.util.toByteArray
import io.scalechain.util.Bytes

object HashCalculator {
  @JvmStatic
  fun transactionHash(tx : Transaction) : Hash {
    val io = CodecInputOutputStream(Unpooled.buffer(), isInput = false)
    TransactionCodec.transcode(io, tx)

    // Run SHA256 twice and reverse bytes.
    assert(io.byteBuf.hasArray())
    val hash = HashFunctions.hash256( io.byteBuf.toByteArray() )

    // BUGBUG : Rethink if using ByteBuf in Hash is a correct apporach.
    return Hash(Bytes(hash.value.array.reversed().toByteArray()))

  }

  @JvmStatic
  fun blockHeaderHash(blockHeader : BlockHeader) : Hash {
    val io = CodecInputOutputStream(Unpooled.buffer(), isInput = false)
    BlockHeaderCodec.transcode(io, blockHeader)

    // Run SHA256 twice and reverse bytes.
    assert(io.byteBuf.hasArray())
    val hash = HashFunctions.hash256( io.byteBuf.toByteArray() )

    return Hash(Bytes(hash.value.array.reversed().toByteArray()))

  }
}


fun Transaction.hash() : Hash {
  return HashCalculator.transactionHash(this);
}

fun BlockHeader.hash() : Hash {
  return HashCalculator.blockHeaderHash(this);
}

/*
object HashSupported {
  implicit fun toHashSupportedBlockHeader(blockHeader: BlockHeader) = HashSupportedBlockHeader(blockHeader)
  implicit fun toHashSupportedTransaction(transaction : Transaction) = HashSupportedTransaction(transaction)
}

data class HashSupportedTransaction(transaction:Transaction)  {
  fun hash() {
    HashCalculator.transactionHash(transaction)
  }
}

data class HashSupportedBlockHeader(blockHeader:BlockHeader)  {
  fun hash() {
    HashCalculator.blockHeaderHash(blockHeader)
  }
}
*/
